src/event/source/unix.rs

```diff
@@ -45,128 +68,128 @@ impl UnixInternalEventSource {
 
-        loop {
-            if let Err(e) = self.poll.poll(&mut self.events, timeout.leftover()) {
-                // Mio will throw an interrupted error in case of cursor position retrieval. We need to retry until it succeeds.
-                // Previous versions of Mio (< 0.7) would automatically retry the poll call if it was interrupted (if EINTR was returned).
-                // https://docs.rs/mio/0.7.0/mio/struct.Poll.html#notes
-                if e.kind() == io::ErrorKind::Interrupted {
-                    continue;
-                } else {
-                    return Err(e);
-                }
-            };
+        #[cfg(feature = "event-stream")]
+        let mut fds = [
+            make_pollfd(&self.tty),
+            make_pollfd(&self.winch_signal_receiver),
+            make_pollfd(&self.wake_pipe.receiver),
+        ];
 
-            if self.events.is_empty() {
-                // No readiness events = timeout
-                return Ok(None);
+        while timeout.leftover().map_or(true, |t| !t.is_zero()) {
+            // check if there are buffered events from the last read
+            if let Some(event) = self.parser.next() {
+                return Ok(Some(event));
             }
-
-            for token in self.events.iter().map(|x| x.token()) {
-                match token {
-                    TTY_TOKEN => {
-                        loop {
-                            match self.tty_fd.read(&mut self.tty_buffer, TTY_BUFFER_SIZE) {
-                                Ok(read_count) => {
-                                    if read_count > 0 {
-                                        self.parser.advance(
-                                            &self.tty_buffer[..read_count],
-                                            read_count == TTY_BUFFER_SIZE,
-                                        );
-                                    }
-                                }
-                                Err(e) => {
-                                    // No more data to read at the moment. We will receive another event
-                                    if e.kind() == io::ErrorKind::WouldBlock {
-                                        break;
-                                    }
-                                    // once more data is available to read.
-                                    else if e.kind() == io::ErrorKind::Interrupted {
-                                        continue;
-                                    }
-                                }
-                            };
-
-                            if let Some(event) = self.parser.next() {
-                                return Ok(Some(event));
-                            }
-                        }
+            match poll(&mut fds, timeout.leftover()) {
+                Err(filedescriptor::Error::Io(e)) => return Err(e),
+                res => res.expect("polling tty"),
+            };
+            if fds[0].events & POLLIN != 0 {
```

Comment:
Should these conditions be checking for `revents` since that is the output indicator of the events that occurred whereas `events` is an input parameter (based on [1] and [2])?

[1] https://docs.rs/filedescriptor/latest/filedescriptor/fn.poll.html
[2] https://man7.org/linux/man-pages/man2/poll.2.html

Reply:
you're right, good catch! The incorrect code works because it just always assumes there is input when the poll returns.